File: src/cpp/LeftRotateString.cpp
class Solution {
public:
    string LeftRotateString(string str, int n) {

        int len = str.size();

        if (len > 0 && n > 0 && n <= len) {
          int firstStart = 0;
          int firstEnd = n - 1;
          int secondStart = n;
          int secondEnd = len - 1;

          string &tmpStr = str;

          // 翻转字符串前面的n个字符
          Reverse(tmpStr, firstStart, firstEnd);
          // 翻转字符串后面剩下的字符
          Reverse(tmpStr, secondStart, secondEnd);
          // 最后翻转整个字符串
          Reverse(tmpStr, firstStart, secondEnd);

        }

        return str;

    }

    void Reverse(string& str, int start, int end) {
      if (start > end || str.size() == 0) {
        return;
      }
      while (start < end) {
        char temp = str[start];
        str[start] = str[end];
        str[end] = temp;

        start++;
        end--;
      }
    }
};

File: src/java/LeftRotateString.java
public class Solution {
    public String LeftRotateString(String str,int n) {
        if (str == null || str.trim().equals("")) {
            return str;
        }
        if (n > str.length()) {
            return "";
        }

        int preStart = 0;
        int preEnd = n - 1;
        int posStart = n;
        int posEnd = str.length() - 1;

        char[] a = str.toCharArray();

        // reverse the previous part
        reverse(a, preStart, preEnd);

        // reverse the post part
        reverse(a, posStart, posEnd);

        // reverse the whole sentence
        reverse(a, preStart, posEnd);

        StringBuffer sb = new StringBuffer(str.length());
        for (char c : a) {
            sb.append(c);
        }

        return sb.toString();
    }

    private void reverse(char[] a, int start, int end) {
        char temp;
        while (start < end) {
            temp = a[start];
            a[start] = a[end];
            a[end] = temp;

            start++;
            end--;
        }
    }
}
